### ISSUE TITLE
Training, Forward / backward pass of tensors with _different_ batch-size, no speedup when backward pass has smaller batch-size

### ISSUE THREAD

**Issue by hfassold** [pytorch/pytorch#74604]
### 🚀 The feature, motivation and pitch

Currently, it seems that in training the Pytorch autograd engine is 'optimized' only for the case where the forward and backward pass are done with tensors with identical batch-size (= last dimension of tensor). If the backward propagation operates on a tensor corresponding to a _smaller_ batch size (e.g. because some samples have been discarded from the batch after the forward pass according to some criterion - e.g. loss value), no speedup is observed, although it _should_ be faster theoretically (because of the smaller batch-size in the backward pass).
This is also mentioned in the NVIDIA GTC 2022 talk 'Faster Neural Network Training, Algorithmically' [1], minute 20:00-21:00
So, although technically it is possible to have a smaller batch size for backward step than in the forward size, currently in Pytorch unfortunately it does not bring a speedup.

 It would be great if pytorch autograd engine could be extended, so that using a smaller batch size in the backward pass leads to a speedup in training. There are several curriculum-learning-like methods which could be utilized then successfully in order to speedup the training, like mini-batch trimming [2] and selective backprop [3]. The basic principle of these methods is to focus more on the 'hard' examples in the batch (the ones with a higher loss - per-sample loss is calculated in forward pass) and discard the 'easier' examples prior to running the backward pass.

--------

[1] NVIDIA GTC 2022, S41421, Faster Neural Network Training, Algorithmically
https://reg.rainfocus.com/flow/nvidia/gtcspring2022/aplive/page/ap/session/1635975950144001iO6H

[2] "Some like it tough: Improving model generalization via progressively increasing the training difficulty", Fassold et al, ISPR, 2022
https://arxiv.org/abs/2110.13058
https://github.com/hfassold/mbtrim

[3] "Accelerating deep learning by focusing on the biggest loser", Jiang et al, 2019
https://arxiv.org/pdf/1910.00762.pdf

**Comment by Lezcano** [comment]
I do not see a-priori why wouldn't you see a speed-up in this case. Can you provide a small self-contained script that tests this?

**Comment by nikitaved** [comment]
Apparently, it is about propagating through full batches in the computation graph even if some of nodes could be pruned..

**Comment by hfassold** [comment]
Good question :-) Actually, both methods (papers [1] and [2], see my original post) are designed to be quite general, could be applied to _any_ model. So a 'general' solution would be great, but from your comments I sense that this is not doable.

An actual _speedup_ with these methods (note the method [1] provides additionally also a better generalization capability) can be only expected  for models _without_ batch-normalization layers. The reason for this is given in [5] (see answer by 'Ivan'). Most CNNs are using batch-normalization layers. I think these could be replaced, without too much harm, by a combination of group normalization and weight standardization, but on the other hand that makes problem when you want to use a pretrained CNN model.

In contrast, transformer architectures used in NLP domain usually do not use batch-normalization, they use layer normalization instead. So I think it makes sense to go for transformer architectures instead in NLP domain. E.g the 'DistillBERT' model is widely used. For this model, I have also a self-contained test code sample, which is not too complicated, at https://github.com/hfassold/nlp_finetuning_adafamily . So I think it could make sense to use the **DistilLBERT** model.

[5] https://stackoverflow.com/questions/68920059/pytorch-no-speedup-when-doing-backward-pass-only-for-a-part-of-the-samples-in-m

**Comment by albanD** [comment]
> I do not see a-priori why wouldn't you see a speed-up in this case.

I think a simple example that shows this is:
```python
out = f(x) # out is very big!
loss = out[0] # My loss only depends on a single entry in out
loss.backward() # This will actually be as expensive as if I used all the elements in out
```

And it is indeed expected and not something that we can easily change. The autograd works at the Tensor level (the whole Tensor) and so we won't be able to see if only a subset of it is being used.
Masked Tensor does sound like the most promising approach here to be able to skip unnecessary operations. But as Christian pointed out, this will potentially require a large number if custom kernels to get to the performance that you expect.
